Stabilize covariance learning with `FillScaleTriL` and update config behaviour

patel-zeel opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current v/s new behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #127

As discussed in #127, the following is the summary of changes:

1. Replace `tfb.FillTriangular` with `tfb.FillScaleTriL` in `config.py`
1. Add the "Custom transformations" section in `examples/uncollapsed_vi.pct.py` to illustrate the use of custom transformation for lower triangular Cholesky parameters. 
1. Change the way config is created currently. Replace `get_defaults()` with `get_global_config()`. Add `reset_global_config()` and `get_default_config()` methods.
1. Add the "Conversion between `.ipynb` and `.py`" section in README.md

### "Custom transformations" section in `examples/uncollapsed_vi.pct.py`
I am showing how to use a custom `triangular_transform` in `gpjax` (the method also generalizes for any transform).

```py
gpx_config = get_global_config()
transformations = gpx_config.transformations
jitter = gpx_config.jitter

triangular_transform = dx.Chain(
    [tfb.FillScaleTriL(diag_bijector=tfb.Square(), diag_shift=jnp.array(jitter))]
)

transformations.update({"triangular_transform": triangular_transform})
```

I have added a point that the `Square` bijector may lead to a faster convergence but can be unstable compared to `Softplus` bijector.

| Softplus | Square |
| :-: | :-: |
| <img width="527" alt="image" src="https://user-images.githubusercontent.com/59758528/208302998-c281cd71-c023-4a5e-806a-988c021bfe66.png">  |  <img width="527" alt="image" src="https://user-images.githubusercontent.com/59758528/208303008-a9def0cd-a259-4bb8-a2a1-e5af891a2293.png">  |


### Config in GPJAX

#### How is it done currently?
1. `get_defaults` method returns the global config. If the config is unavailable, it creates one.

#### What is the new behavior?
1. `get_global_config` returns the global config. If the config is unavailable, it creates one. If JAX precision changes, it makes appropriate changes to the current global config e.g. update `FillScaleTriL` 
2. `get_default_config` returns the default config and does not update any global config.
3. `reset_global_config` resets global config to default config.

### Conversion between `.ipynb` and `.py`
The following quick commands are introduced in README.md to convert between `.ipynb` and `.py`.
![image](https://user-images.githubusercontent.com/59758528/208302251-8be6692d-093f-49ca-ba12-10aa9b2cfe7c.png)

> ### Some changes are attempted in this PR without a detailed prior discussion. Please consider those as suggestions and suggest your thoughts on the required modifications.

## What is not done?

I have not ensured yet that `get_global_config` does not get invoked at `import gpjax as gpx` (O1 in #127). Execution of https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/__init__.py#L16 runs [natural_gradients.py](https://github.com/JaxGaussianProcesses/GPJax/blob/master/gpjax/natural_gradients.py) which invokes `get_global_config`. Any suggestions on how to solve this?

Daniel-Dodd commented:
Thanks, @patel-zeel! This PR looks great. Will start my review soon. 

> ## What is not done?
> I have not ensured yet that `get_global_config` does not get invoked at `import gpjax as gpx` (O1 in #127). Execution of
> 
> https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/__init__.py#L16
> 
> runs [natural_gradients.py](https://github.com/JaxGaussianProcesses/GPJax/blob/master/gpjax/natural_gradients.py) which invokes `get_global_config`. Any suggestions on how to solve this?

Is `get_global_config` invoked in `natural_gradients.py` solely via `DEFAULT_JITTER`? We can then move the jitter inside the relevant functions/objects as done in `gps.py` and `variational_families.py`, to resolve this.

Daniel-Dodd reviewed (changes requested):
Thanks @patel-zeel, this is an excellent PR. Just a few minor things, and to see if O1 in https://github.com/JaxGaussianProcesses/GPJax/issues/127 might be resolved via removing the default jitter and define jitter inside objects/functions akin to `gps.py`/`variational_families.py`.

patel-zeel commented on examples/uncollapsed_vi.pct.py:
```diff
@@ -191,9 +198,61 @@
+# %% [markdown]
+# ## Custom transformations
+#
+# To train a covariance matrix, `gpjax` uses `tfb.FillScaleTriL` transformation by default. `tfb.FillScaleTriL` fills a 1d vector into a lower triangular matrix and then applies `SoftPlus` transformation on the diagonal to satisfy the necessary conditions for a valid Cholesky matrix. Users can change this default transformation with another valid transformation of their choice. For example, `Squa ...
```
Done.

patel-zeel commented on examples/uncollapsed_vi.pct.py:
```diff
@@ -191,9 +198,61 @@
+
+# %% [markdown]
+# We can see that `Square` transformation is able to get relatively better fit compared to `Softplus` with the same number of iterations, but `Softplus` is recommended over `Square` for stability of optimization.
+
```
Done.

patel-zeel commented on gpjax/config.py:
```diff
@@ -21,35 +22,88 @@
+        [
+            tfb.FillScaleTriL(diag_shift=jnp.array(__config.jitter)),
+        ]
+    )  # TODO: Dan to chain methods.
```
Done.

patel-zeel commented on README.md:
````diff
@@ -48,6 +48,20 @@ Feel free to join our [Slack Channel](https://join.slack.com/t/gpjax/shared_invi
 > - [**Custom kernels**](https://gpjax.readthedocs.io/en/latest/examples/kernels.html#Custom-Kernel)
 > - [**UCI regression**](https://gpjax.readthedocs.io/en/latest/examples/yacht.html)
 
+## Conversion between `.ipynb` and `.py`
+Above examples are stored in [examples](examples) directory in the double percent (`py:percent`) format. Checkout [jupytext using-cli](https://jupytext.readthedocs.io/en/latest/using-cli.html) for more info.
+
+* To convert `example.py` to `example.ipynb`, run:
+
+```bash
+jupytext --to notebook example.py
+```
+
+* To convert `example.ipynb` to `example.py`, run:
+
+```bash
+jupytext --to py:percent example.ipynb
+```
````
Thank you!

patel-zeel commented on gpjax/config.py:
```diff
@@ -21,35 +22,88 @@
+    # If the global config is available, check if the x64 state has changed
+    x64_state = jax.config.x64_enabled
+
+    # If the x64 state has not changed, return the existing global config
+    if x64_state is __config.x64_state:
+        return __config
+
+    # If the x64 state has changed, return the updated global config
+    update_x64_sensitive_settings()
+    return __config
```
Thanks!

patel-zeel commented:
Thank you for the review, @Daniel-Dodd. I have resolved O1 in #127 by moving the jitter inside the function. I have also addressed the other comments.

patel-zeel commented:
@Daniel-Dodd since we have addressed O1 in #127, we may need to ensure it does not get violated in the future. I think it is hard to detect such things via unit testing. However, I have added a relevant test (`test_config_on_library_import`) in `test_config` which calls `get_global_config_if_exists` function to check if the global config exists. This test seems to be passing only because of two reasons: i) `test_config_on_library_import` test is defined on top and thus it is executed first within `test_config.py` ii) All tests running before `test_config.py` do not invoke the creation of global config. What are your thoughts on this?

Daniel-Dodd commented:
Hi @patel-zeel, thank you for addressing my comments.

> @Daniel-Dodd since we have addressed O1 in #127, we may need to ensure it does not get violated in the future.

In agreement with this. However, the latest test workflow failed on `test_config_on_library_import` (https://github.com/JaxGaussianProcesses/GPJax/actions/runs/3729043958/jobs/6330357412)? Happy to merge this PR with this test removed for now, and open a separate issue regarding checks for O1 violation?

patel-zeel commented:
Yes, that looks like a way to go :)

Daniel-Dodd commented:
@patel-zeel Nice! Running the test workflow now, will merge soon as the tests pass. :)
